src/test/java/com/artipie/gem/GemITCase.java

```diff
@@ -0,0 +1,160 @@
+ * SOFTWARE.
+ */
+package com.artipie.gem;
+
+import com.artipie.ArtipieServer;
+import com.artipie.RepoConfigYaml;
+import com.artipie.RepoPerms;
+import com.artipie.asto.Key;
+import com.artipie.asto.Storage;
+import com.artipie.asto.fs.FileStorage;
+import com.artipie.asto.test.TestResource;
+import com.artipie.test.RepositoryUrl;
+import com.artipie.test.TestContainer;
+import java.io.IOException;
+import java.nio.file.Path;
+import org.cactoos.text.Base64Encoded;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.core.IsEqual;
+import org.hamcrest.core.StringContains;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.EnabledOnOs;
+import org.junit.jupiter.api.condition.OS;
+import org.junit.jupiter.api.io.TempDir;
+
+/**
+ * Integration tests for Gem repository.
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ * @since 0.13
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+@EnabledOnOs({OS.LINUX, OS.MAC})
+final class GemITCase {
+
+    /**
+     * Repo.
+     */
+    private static final String REPO = "my-gem";
+
+    /**
+     * Rails gem.
+     */
+    private static final String RAILS = "rails-6.0.2.2.gem";
+
+    /**
+     * Temporary directory for all tests.
+     * @checkstyle VisibilityModifierCheck (3 lines)
+     */
+    @TempDir
+    Path tmp;
+
+    /**
+     * Tested Artipie server.
+     */
+    private ArtipieServer server;
+
+    /**
+     * Container.
+     */
+    private TestContainer cntn;
+
+    /**
+     * Repository url.
+     */
+    private RepositoryUrl url;
+
+    /**
+     * Storage.
+     */
+    private Storage storage;
+
+    @Test
+    public void gemPushWorks() throws Exception {
+        final boolean anonymous = true;
+        this.init(anonymous);
+        this.push(anonymous);
+        MatcherAssert.assertThat(
+            this.storage.exists(
+                new Key.From("repos", GemITCase.REPO, "gems", GemITCase.RAILS)
+            ).toCompletableFuture().join(),
+            new IsEqual<>(true)
+        );
+    }
+
+    @Test
+    void gemInstallPushedGemWorks() throws Exception {
+        final boolean anonymous = true;
+        this.init(anonymous);
+        this.push(anonymous);
+        MatcherAssert.assertThat(
+            this.cntn.execStdout("gem", "install", GemITCase.RAILS, "--ignore-dependencies"),
+            new StringContains("Successfully installed rails-6.0.2.21")
+        );
+    }
+
+    @AfterEach
+    void tearDown() {
+        this.server.stop();
+        this.cntn.close();
+    }
+
+    private void push(final boolean anonymous) throws Exception {
+        final String key = new Base64Encoded("usr:pwd").asString();
+        new TestResource(String.format("gem/%s", GemITCase.RAILS))
+            .saveTo(this.storage, new Key.From(GemITCase.RAILS));
+        final String urlwithslash = this.url.string(anonymous);
```

Comment:
@genryxy the name does not make much sense, why URL is with slash? Maybe `urlwithuser` or just `url` would be better?

Reply:
@olegmoz in this case I want to show that `this.url.string(anonymous)` returns url with slash and request looks like `POST /my-gem//api/v1/gems HTTP_1_1`. So, I take substring of `urlwithslash`

Reply:
@genryxy but `this.url` contains these slashes. How calling `string` method on `this.url` transforms it into URL with slashes (adds slashes)?

Reply:
@olegmoz `string` method does not add slash at the end. It just concatenates as `String.format("http://%s", this.hostportrepo)`. `hostportrepo` is initialized in the constructor
```
public RepositoryUrl(final int port, final String reponame) {
    this.hostportrepo = String.format("host.testcontainers.internal:%d/%s/", port, reponame);
}
```

Reply:
@genryxy exactly my point, it does not add slashes. Then why code reads "apply method `string` to `this.url` with parameter `anonymous` and get URL with slashes as result`, does it make sense?

Reply:
@olegmoz thanks, renamed variable